### PR TITLE
Building for Android by Android NDK.

### PR DESCRIPTION
**Pull request by alexeikh-promon**
Support for building OpenSSL for four 32-bit architectures supported by Android NDK.

Included:
- Changes to "config" and "Configure" scripts.
- New file: setenv-android.sh script, based on
  http://wiki.openssl.org/images/7/70/Setenv-android.sh script, mentioned at
  http://wiki.openssl.org/index.php/Android .

The compiler options in "Configure" script are based on:
- Existing options from previous versions of this script.
- Options that are supplied by Android NDK r10 for linux-x86_64, toolchains:
  arm-linux-androideabi-4.6, mipsel-linux-android-4.6, x86-4.6.
  As can be seen, NDK supplies quite different options are for different
  architectures.
- Considerations about footprint vs speed of the resulting code on mobile
  devices.

### CODE REVIEW

**Comment by alexeikh-promon** [comment]
Hi, Patch Reviewers.

This pull request is in queue for almost 4 months already. I'd like to note that:

* The main part of the patch ("Configure" and "config" files) is short and simple.
* The added set-env script is based on the script that already resides on OpenSSL Wiki.
* Patches from me were accepted before (in pre-github era), for what it's worth.

Can I please get some attention to my pull request?

**Comment by richsalz** [comment]
in general RT is the place we look first. four months isn't long :) we're working on improving but ...

**Comment by richsalz** [comment]
@levitte isn't this done now as part of the config revamp?

**Comment by levitte** [comment]
Uhmmmm, not at all sure

**Comment by FloridaStream** [comment]
Hi guys, you can add support to generate `objs` files in `/obj/local/<CPU_ABI>/objs/` `crypto/`, `openssl/`, `ssl/` and `ssltest/`
Sample as generated by https://github.com/yixia/OpenSSL-Vitamio

I only get this files in ``/obj/local/<CPU_ABI>/` `libcrypto.so` and `libssl.so`
Thank You :+1:

**Comment by mattcaswell** [comment]
Not sure what we do with this one, but my guess is part of the problem is no one on the dev team has access to this platform to test this out.

**Comment by alexeikh-promon** [comment]
I'd like to leave some thoughts here.

1) During 19 months lifetime of this pull request, nobody found an issue with it, nor proposed a better patch. I don't know how many people on the Internets tested the patch, but this patch works for our company for all those 19 months.

2) "my guess is ... no one on the dev team has access to this platform". It is pretty easy to get access to the Android platform. One can buy an Android phone (or download an emulator!), compile OpenSSL using this patch, write a sample program or get one of the existing OpenSSL tools compiling, and run that program on the phone. It can be a pure-C command-line program, compiled by standalone NDK toolchain and be ran from "adb shell". You don't have to make an Android project with Java, XML, Manifest, Activities or JNI in order to test.

3) I see support for many rare platforms in the "Configure" script, much rarer than Android. Does it mean that every platform was actually tested by an OpenSSL dev team member? I doubt it.

4) What can go wrong? What is the worst thing that can happen if the patch gets included? Going from "unsupported" state to "badly supported"? Hidden exploit? Look at the patch, the only changes in "Configure" and "config" scripts are compilation flags.  No C code is changed.

5) Google is regularly releasing newer Android NDK versions and sometimes even adding new architectures (64-bit ones). If my patch gets merged - I will feel more motivated to update the patch for newer NDKs. FWIW.

**Comment by bmurray7** [comment]
Clover Network is using this patch to deploy OpenSSL on dev devices. We have found no issues.

**Comment by kroeckx** [comment]
This clearly doesn't apply to 1.1.0 or master.

**Comment by alexeikh-promon** [comment]
Sure it doesn't, after 2+ years. I am willing to update the patch, but not sure how to do it. GitHub says "Add more commits by pushing to the building-for-android branch on alexeikh-promon/openssl." Just adding commits to the existing branch will not help, because the current branch is branched off too old commit. What to do then? Rebase the branch to the latest master? Or make a new branch and new pull request?

**Comment by kroeckx** [comment]
Rebasing and forcing a push should do it.

**Comment by yan12125** [comment]
An idea: NDK now recommends clang over gcc, and Google is planning drop gcc. It might be better to add clang support instead of fix gcc.

FYI: I have a patch that uses clang and unified headers introduced in NDK r14 beta1 at https://github.com/yan12125/python3-android/blob/master/mk/openssl/fix-cflags.patch.

**Comment by alexeikh-promon** [comment]
Thanks for the help, kroecx.

Build scripts on OpenSSL_1_1_0-stable or master are indeed very different from OpenSSL_1_0_2-stable branch or 2-years-old master. I don't know when I will have time to rework this patch for master.


In the meantime, for those who still use OpenSSL 1.0.2 or earlier, I've updated this patch to the tip of OpenSSL_1_0_2-stable branch. The new patch can be found on this branch:

https://github.com/alexeikh-promon/openssl/commits/OpenSSL_1_0_2-stable.Building_for_Android.2017-01

The new patch consists of several commits, starting from commit edcc244 "Remove generated assembly files during "make clean"."


yan12125, your idea totally makes sense. I tried to introduce clang and unified headers support (still on OpenSSL_1_0_2-stable, but it wasn't very easy). Particularly, when trying clang instead of gcc, the build kept failed on "makedepend" and it wasn't obvious to me how to fix it. Unified headers support is currently only available in NDK r14 beta 1, but this NDK version somewhy ignores NDK_TOOLCHAIN_VERSION parameter (though respects APP_UNIFIED_HEADERS parameter) and forces clang even if I ask for gcc. So it was a bit difficult to try unified headers on that branch too. Well, unified headers for an older OpenSSL branch are not so relevant.

**Comment by petrovr** [comment]
Build with clang is not in scope of this issue and all my following comments are only for gcc.

I cannot see any reason to update build configuration for 1.0* branches. 
The current configure targets are enough to me to build OpenSSL for all six supported by Android NDK (abiFilters) : armeabi, armeabi-v7a, x86, mips, arm64-v8a, x86_64 and mips64.

1) "android" is generic(!) configure options and must not be touched. It could be used for 64-android builds. 

2) OpenSSL builds system allow  compiler or linked flags to be passed on command line. As for other openssl build system for android Configure should contain minimum of flags. 
It is end user responsibility to choose those flags. For instance build engineer may decide to use "hardening" flags.

3) It is user responsibility to define environment variables that help him easy to switch(upgrade) to another ndk.

In conclusion:
Modification that remove "android" from Configure is wrong. 
Rest of the patch belongs to build guide. 


I would like to propose issue to be closed.

**Comment by yan12125** [comment]
Here I'd like to proposed my pull request at #2229. It implements GCC & Clang dual support with Android NDK. Unified headers support is not included - I'd like to postpone it until NDK r14 stable is released.

> Particularly, when trying clang instead of gcc, the build kept failed on "makedepend"

I guess either the latest OpenSSL or the latest NDK has fixed it.

> Unified headers support is currently only available in NDK r14 beta 1, but this NDK version somewhy ignores NDK_TOOLCHAIN_VERSION parameter (though respects APP_UNIFIED_HEADERS parameter) and forces clang even if I ask for gcc.

Sounds an NDK bug. Anyway OpenSSL doesn't use ndk-build or Android.mk at all, so those problems are irrelevant.

**Comment by richsalz** [comment]
This is outdated; see also #2229

**Comment by neerajgoel** [comment]
Hi all, i want to make a simple SSL connection with server in android NDK.  **Could anyone tell me how to setup SSL lib in android NDK step by step?**. Any help would be appreciated.
